test/unit/remote-config/remote-config-api-client.spec.ts

```diff
@@ -0,0 +1,161 @@
+    error: {
+      code: 404,
+      message: 'Requested entity not found',
+      status: 'NOT_FOUND',
+    },
+  };
+  const EXPECTED_HEADERS = {
+    'Authorization': 'Bearer mock-token',
+    'X-Firebase-Client': 'fire-admin-node/<XXX_SDK_VERSION_XXX>',
+    'Accept-Encoding': 'gzip',
+  };
+  const noProjectId = 'Failed to determine project ID. Initialize the SDK with service '
+    + 'account credentials, or set project ID as an app option. Alternatively, set the '
+    + 'GOOGLE_CLOUD_PROJECT environment variable.';
+
+  const mockOptions = {
+    credential: new mocks.MockCredential(),
+    projectId: 'test-project',
+  };
+
+  const clientWithoutProjectId = new RemoteConfigApiClient(
+    mocks.mockCredentialApp());
+
+  // Stubs used to simulate underlying api calls.
+  let stubs: sinon.SinonStub[] = [];
+  let app: FirebaseApp;
+  let apiClient: RemoteConfigApiClient;
+
+  beforeEach(() => {
+    app = mocks.appWithOptions(mockOptions);
+    apiClient = new RemoteConfigApiClient(app);
+  });
+
+  afterEach(() => {
+    _.forEach(stubs, (stub) => stub.restore());
+    stubs = [];
+    return app.delete();
+  });
+
+  describe('Constructor', () => {
+    it('should throw when the app is null', () => {
+      expect(() => new RemoteConfigApiClient(null as unknown as FirebaseApp))
+        .to.throw('First argument passed to admin.remoteConfig() must be a valid Firebase app instance.');
+    });
+  });
+
+  describe('getTemplate', () => {
+    it(`should reject when project id is not available`, () => {
+      return clientWithoutProjectId.getTemplate()
+        .should.eventually.be.rejectedWith(noProjectId);
+    });
+
+    it('should resolve with the requested template on success', () => {
+      const stub = sinon
+        .stub(HttpClient.prototype, 'send')
+        .resolves(utils.responseFrom({
+          conditions: [{ name: 'ios', expression: 'exp' }],
+          parameters: { param: { defaultValue: { value: 'true' } } },
+          version: {},
+        }, 200, { etag: 'etag-123456789012-1' }));
+      stubs.push(stub);
+      return apiClient.getTemplate()
+        .then((resp) => {
+          expect(resp.conditions).to.deep.equal([{ name: 'ios', expression: 'exp' }]);
+          expect(resp.parameters).to.deep.equal({ param: { defaultValue: { value: 'true' } } });
+          expect(stub).to.have.been.calledOnce.and.calledWith({
+            method: 'GET',
+            url: 'https://firebaseremoteconfig.googleapis.com/v1/projects/test-project/remoteConfig',
+            headers: EXPECTED_HEADERS,
+          });
+        });
+    });
+
+    it('should throw when the etag does not present in the headers', () => {
+      const stub = sinon
+        .stub(HttpClient.prototype, 'send')
+        .resolves(utils.responseFrom({
+          conditions: [], parameters: {}, version: {},
+        }));
+      stubs.push(stub);
+      const expected = new FirebaseRemoteConfigError('invalid-argument', 'ETag is unavailable');
```

Comment:
Make this error message a bit more descriptive. `ETag header not present in server response.`